README
#######################################################################
# AUT UpStage Team 2011
# Author: Heath Behrens (2011)
#
# ChangeLog:
# Updated Dependencies (Heath Behrens 13-04-2011)
# Updated Instructions and added note on dependencies. (Heath Behrens 05-05-2011)
# Updated Compiling of the client.
#
#######################################################################

List of dependencies

#######################################################################

espeak 
festival 
gif2png 
libgif4 
netpbm 
python-twisted <= 8.1.0-4
lame >= 3.97-0.0
libgdbmg >= 1.7.3-28
rsynth >= 2.0-6
mbrola >= 3.01h-6
swftools >= 0.9.0-0ubuntu1
python <= 2.5.2

#######################################################################

Install Instructions:

As root run: 

1) apt-get install espeak festival gif2png libgif4 netpbm python-twisted timeout

If you have downloaded the other dependencies as deb files, run:

2) Extract the dependency.tar.gz file 

3) Cd to the dependency folder and run as root:
	dpkg -i *.deb 

-NOTE. You will need to install: swftools, lame, libgdbmg, mbrola and rsynth using your distributions package manager.
 
 swftools: wget http://archive.canonical.com/ubuntu/pool/partner/s/swftools/swftools_0.9.0-0ubuntu2_i386.deb
 rsynth: wget http://archive.debian.org/debian-archive/debian/pool/non-free/r/rsynth/rsynth_2.0-6_i386.deb

If Pulled from svn skip to step 5.

4) Extract UpStage-2.x.x.zip or .tar.gz

5) open a terminal and cd to 2.x.x directory.
    Run:
	    python install.py (NOTE this will install upstage. Refer to below instructions for other options)
	
If you prefer to use the deb file there is a built in deb creater but you need to have the following installed:

apt-get install build-essential autoconf automake autotools-dev dh-make debhelper devscripts fakeroot xutils lintian pbuilder

Once the above build tools are installed run: 
	'python install.py deb' (without quotes to create a deb package)

To compile the client if you made changes or just want to make sure the latest client in included, use:
	python install.py cc /path/to/flex
If you do not have the flex compiler 2. Then this will not work.

Changelog
Changelog

30-04-2011 (Heath & Henry)
-Made several changes to syntax.
-Added missing functions and configuration information.

04-05-2011 (Heath, Vibhu, Henry)
-Fixed Login bug (74-75) Logs in correctly first attempt on all browsers.
05-05-2011 (Vibhu, Heath)
-Added new redirect to home page.

06-05-2011 (Heath, Vibhu)
-Fixed text-to-speech audio bug so text-to-speech works and the sound slots get re-assigned.

08-05-2011 (Heath)
-Made some alterations to the install.py namely fixed a bug which caused the script to fail when generating the deb file.
-The install.py now removes all .svn files till I can figure out how to ignore them. Also fixed the bug with permissions for the debian package scripts.
-The install.py script can now compile the client if the flex compiler is present on the users system. REFER to README file for instructions.

17-05-2011 (Heath, Moh, Vibhu)
-Fixed media filter. Now correctly filters media and all names remain in tact.
-Media upload works correctly in all browsers (including firefox 4).

18-05-2011 (Heath, Moh)
-Fixed change own password, now correctly changes own password(server side).
-Added checks to user.js which ensure an empty password cannot be entered.

20-05-2011 (Heath)
-Fixed Bubble now fades after some time which is calculated based on the length of the message.
-Fixed the Bubble is removed from the stage when the avatar is dropped.

07-06-2011 (Heath & Mohammed)
-Fix for duplicate footer in pages stageedit and media edit. Edited file masterpage.js function fillpage line 297.

12-06-2011(Heath & Mohammed)
-Moved the code to remove duplicated footer into stageEdit() in stageedit.js. Also added the version info
in master_b.inc to the div.

13-06-2011(Vibhu, Moh, Heath, Henry, JR)
-Various html layout bugs fixed, the entire website has been revamped using a new template and layout.
-Fixed image scrolling. Images are now spaced correctly and names do not overlap. Improved look and feel.
-Fixed server responses which where breaking html layout within tables.
-Changed versioning information to reflect current version (2.4.1).

15-06-2011 (Vibhu, Heath & Mohammed)
-Backdrop resizing, so it remains only in stage
-Centered the stage display
-Backdrop remains behind avatar

17-06-2011 (Heath Behrens)
-Fixed updating voices, changing a voice will not require a server restart.
-Fixed bubble fading, the alpha value now slowly decreases (this can be modified very easily to fade slower or faster, for now its just a static fade duration).

18/06/2011 (Heath Behrens)
-Increased width of speech bubble and also slightly modified the duration of the speech bubble on stage, to further take into account the length of the text.
-Fixed 2 errors that prevented audio files from being selected and tested on the mediaedit page. Changes namely made to pages.py. Also removed another log.msg that printed the bytes of audio files to the log file (Twice!).

26/06/2011 (Heath)
-Fixed error on speech bubble timer which threw an exception, due to not stopping all timers correctly.

05/07/2011 (Heath Behrens)
-Fix for exception loading audio files, namely null objects. Still no music playback though.
-Removed the icon from Audio.as as a temporary fix for it covering the play buttons.

28/07/2011 (Heath Behrens / Vibhu Patel)
-Merged 2.1 client (stage) back into 2.4.1 codebase.
-removed .pyproject files.
-Fixed an error when uploading media with more than 1 . character in the name.
-Changed the layout of stageedit page and removed previous buttons replacing them with radio buttons.
-Switched assign to stages boxes on the media upload page.

08/08/2011 (Heath Behrens / Vibhu Patel)
-Fixed Avatar scaling in client, avatars now scale correctly in Actionscript 2 client.
-Also added the actionscript 2 compile script to the source as it is required.
-On a side note the AS3.client can be renamed to client. In which case the actionscript 3 client will be used instead of AS2 client.
-Fixed prop so it scales on stage.

09/08/2011 (Heath Behrens & Vibhu Patel)
-Merged avatar menu into 2.1 all functions working.
-Fixed modifying colours in media managment system namely changing the background colour of chat,prop etc.
The fix is part of an html issue which only changed the background colour of the prop bar. 
-Radio buttons removed from the stage edit page, instead images are used as it looks more flashy and the requests are sent correctly.
-compileswf was modified to use the current directory opposed to the home directory.

10/08/2011 (Heath Behrens)
-Fixed parsing of xml file on server, now the stage access options options are extracted correctly. 
 However the stageedit page needs some tweaking.

12/08/2011 (Heath Behrens / Vibhu Patel)
-Reworked retrieving media info on the server side, this was part of a fix which did not show unassigned stages in the workshop. Instead all stages where shown as assigned even if that was not the case.
-Added session tracking to masterpage.js so a user stays logged in for 24hrs provided cookies are allowed with their browser.

16/08/2011 (Heath Behrens / Vibhu Patel)
-Completed fixes for assign and unassign stages, so all is working. 
-Added a patch to the request sent in masterpage to correctly send the request to the server for assigning and unassigning stages.
-Audience and player counters for each stage are displayed on each web page.

19/08/2011 (Heath Behrens / Vibhu Patel)
-Filtering optimized and allows multiple filters of the same type. There are two different types of filtering are available. Namely media can be filtering by removing items not matching the filtering criteria. The other method includes all media matching only the filter options.
-Fixed media scrolling the position is remebered an returned to when selecting an item.
-Layout of media edit page works cross all browsers (tested browsers).

upstage/things.py
#Copyright (C) 2003-2006 Douglas Bagnall (douglas * paradise-net-nz)
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

"""
Things that go on the stage:

  Thing    - base class for things
  Avatar   - actors on the stage
  Prop     - other objects on stage
  Backdrop - big props in front of which stuff happens.
  Audio    - playable music and sound effects in the audio widget
  
Author: 
Modified by: Endre Bernhardt, Phillip Quinlan
Modified by: 18/05/11 Mohammed Al-Timimi and Heath Behrens - fixed get_uploader_list
Modified by: Heath Behrens (17/06/2011) - Added a helper method to check if the ThingCollection
             contains a given thing. Part of the updating voices fix.
Modified by: Heath Behrens 16/08/2011 - Added functions remove_media and remove_mediafile which are called
                                        to remove media from the stage.
             Karena, Corey, Heath 24/08/2011 - Added tag variable to Thing class to store tags for a given thing.          
Notes: 
"""

#siblings
from upstage.util import id_generator
from upstage.misc import UpstageError
from upstage import config


#twisted
from twisted.python import log

## @brief _next_thing_id id_generator for things
_next_thing_id = id_generator(start=config.THING_MIN_ID,
                              prefix='').next

_nullpos = (None,None,None)


class Thing:
    """Base representation of a thing on stage"""
    def __init__(self, media, name="", position=_nullpos, ID=None):
        """
        @param media -- id of media file
        @param name name of the thing
        @param position x, y and z position
        """
        if ID is None:
            self.ID = _next_thing_id()
        else:
            self.ID = ID
        self.name = name or media.name
        self.media = media
        self.X, self.Y, self.Z = position
        #Karena, Corey, Heath 24/08/2011 - Added for media tagging. Stores tag in each thing
        self.tags = ''

    def get_pos(self):
        """Retrieve tuple of x, y, z coordinates"""
        return (self.X, self.Y, self.Z)

    def move_to_layer(self, newlayer):
        self.layer = newlayer

    def move(self,X,Y,Z=None):
        """Change position.  Really need some way of representing slow
        motion
        
        @param X x position
        @param Y y position
        @param Z z position"""
        (self.X, self.Y) = X,Y
        if Z is not None:
            self.Z = Z #XXX no way of setting Z back to None. does it matter? (Z isn't *used*)

    def exit(self):
        """Set position to Nowhere"""
        self.X = None
        self.Y = None
        self.Z = None



class Avatar(Thing):
    """Representation of a moving talking thing on stage"""    
    typename= 'avatar'

    def __init__(self, media, name="", position=_nullpos,
                 voice=None, show_name='show', ID=None):
        """
        @param media:      a globalmedia.mediafile
        @param name:       avatar name
        @param position:   x, y, z position of avatar
        @param voice:      voice of the avatar
        @param show_name: 'show' or 'hide'
        """        
        Thing.__init__(self, media, name=name, position=position, ID=ID)
        self.socket = None
        if not voice and hasattr(media, 'voice'):
            self.voice = media.voice
        else:
            self.voice = voice
        self.layer = 10000 + (int(self.ID) * 10)
        self.show_name = show_name
        self.prop = None
        self.frame = 1

    def set_frame(self, frame):
        self.frame = frame;

    def hold_prop(self, prop):
        """hold a prop"""
        #recorded twice, circularly, on the prop and on the avatar.
        # Allow stealing of prop from other avatars
        if prop.holder is not None:
            prop.holder.drop_prop()

        self.prop = prop
        prop.holder = self

    def drop_prop(self):
        """drop the help prop"""
        self.prop.holder = None
        self.prop = None

    def allows_player(self, player):
        """hook for player rights."""
        return True #XXX unused, deprecated.


class Prop(Thing):
    """Representation of a still thing on stage.
    """
    # not too different from a generic Thing
    typename= 'prop'
    load_command = 'LOAD_PROP'

    def __init__(self, media, name="", position=_nullpos, ID=None):
        Thing.__init__(self, media=media, name=name, position=position, ID=ID)
        self.holder = None



class Backdrop(Thing):
    """Representation of a big still thing at the back of the stage"""
    # not too different from a Prop
    typename= 'backdrop'
    load_command = 'LOAD_BACKDROP'
    frame = 1;
    def __init__(self, media, name="", position=_nullpos, ID=None):
        Thing.__init__(self, media=media, name=name, position=position, ID=ID)

    def set_frame(self, frame):
        self.frame = frame;
        log.msg('Backdrop ID %s is updated to frame: %s' %(self.ID, self.frame))


# PQ & EB: Created Audio - 17.9.07
class Audio(Thing):
    """Representation of an audio file"""
    typename= 'audio'
    def __init__(self, media, displayName="", position=_nullpos, ID=None):
        Thing.__init__(self, media=media, name=displayName, position=position, ID=ID)



#
#~~~~~~~~~~~~~~~~~~~~~~~~~~ ThingCollection ~~~~~~~~~~~~~~~~~~


class ThingCollection:
    """Stores a collection of things for a stage.  It indexes the
    collection in two ways: on the members 'ID' and 'media'
    attributes.
    """

    def __init__(self, Class, globalmedia, idfunc=None):
        """The collection will contain things
        belonging to the passed in class.

        @param Class: the Thing subclass that this is a collection of.
        @param idfunc: a function generating unique IDs. If None, a
                       new generator will be formed.
        """
        self.things = {}
        self.media = {}        
        self.Class = Class        
        self.globalmedia = globalmedia
        self.typename = Class.typename
        self._next_thing_id = idfunc or id_generator(start=config.THING_MIN_ID,
                                                     prefix='').next


    def add_media(self, media):
        """Add a media item to the collection.  If the item already
        exists, the old details for it are over written.  Returns the
        instance of the Thing subclass created and stored here, or
        None if won't work.

        @param media: a globalmedia.MediaFile instance belonging to the
                      correct global media dictionary.
        """
        f = media.file
        if f not in self.globalmedia:
            raise ValueError("rejecting %s, not found in globalmedia" % f)

        if f in self.media:   #delete from both dictionaries.
            oldthing = self.media.pop(f)
            del self.things[oldthing.ID]

        ID = self._next_thing_id()
        #create the Thing instance
        thing = self.Class(media, ID=ID)
        
        if (self.Class == Audio):
            thing.type = media.medium
    
        self.things[ID] = thing
        self.media[f] = thing
        return thing

    """

    Heath Behrens / Vibhu Patel 16/08/2011
        -Fucntion used to remove media from the Thing collection

    """    
    def remove_media(self, media):
        f = media.file #get reference to the media file
        if f not in self.globalmedia:
            raise ValueError("rejecting %s, not found in globalmedia" % f)
        if f in self.media:   #make sure to delete from both dics
            oldthing = self.media.pop(f)
            del self.things[oldthing.ID]

    """
     Function to add a media file to the thing collection
    """
    def add_mediafile(self, f):
        """try to find the filename f in self.globalmedia.
        If successful, add the corresponding media object,
        returning the approriate Thing"""
        try:
            media = self.globalmedia[f]
        except KeyError:
            raise UpstageError('no such media as %s in %s thing collection' % (f, self.typename))
        return self.add_media(media)

    """

    Heath Behrens / Vibhu Patel 16/08/2011
        -Used to remove a media file from the collection, essentially a callback function

    """
    def remove_mediafile(self, f):
        try:
            media = self.globalmedia[f]
        except KeyError:
            raise UpstageError('Media file does not exist in thing collection')
        self.remove_media(media)  
              
    """
        Added By: Heath Behrens 17/06/2011 - Simple helper method that checks if this collection contains
        the given thing.
        Returning true if so and false otherwise
    """
    def contains_thing(self, thing):
        f = thing.file
        if f in self.media:
            return True # found in collection
        return False # not found
                
    def drop_mediafile(self, f):
        """get rid of the thing indicated to by media filename f"""
        thing = self.media.pop(f, None)
        
        if thing is not None:
            del self.things[thing.ID]
        else:
            log.msg('drop_mediafile: this is None - item wasn\'t deleted')

    def clear(self):
        """Delete all media in dictionary"""
        self.things = {}
        self.media = {}        

    def __str__(self):
        return "<ThingCollection containing %d things of class %s>" %\
               (len(self.things), self.Class)

    def html_list(self, uploader='', include='thing_item_set.inc'):
        """make a list of things as a set of table rows,
        with tick boxes indicating whether thery are selected (etc)"""
        selected = self.media.keys()
        return self.globalmedia.html_list(uploader, include, selected, prefix='')
    
    """ 18/05/11 Mohammed + Heath - Get uploader list """
    def get_uploader_list(self, include='uploader_item_set.inc'):
        """ makes a collection of all the unqiue uploaders for the current media type """
        uploaders = self.globalmedia.get_formatted_list() 
        return uploaders
    
	
    def __contains__(self, ID):
        return ID in self.things

    def __getitem__(self, ID):
        return self.things[ID]

    def get(self, ID, default=None):
        return self.things.get(ID, default)

    def reap_zombies(self):
        """If a thing refers to media not in the globalmedia superset,
        drop it.  Return True if successful, otherwise False"""
        found = False
        for f in self.media.keys():
            if f not in self.globalmedia:
                thing = self.media.pop(f)
                del self.things[thing.ID]
                found = True
        return found